### DISCUSSION TITLE
[Question]: About PAYLOAD GUIDs

### DISCUSSION THREAD

**Discussion by lshqqytiger** [vosen/ZLUDA#84] [Q&A] [answered]
First of all, I'm not an expert of low-level programming.
I saw these guids in `zluda_redirect\src\payload_guid.rs`
```rust
#[allow(dead_code)]
const PAYLOAD_NVCUDA_GUID: detours_sys::GUID = detours_sys::GUID {
    Data1: 0xC225FC0C,
    Data2: 0x00D7,
    Data3: 0x40B8,
    Data4: [0x93, 0x5A, 0x7E, 0x34, 0x2A, 0x93, 0x44, 0xC1],
};

#[allow(dead_code)]
const PAYLOAD_NVML_GUID: detours_sys::GUID = detours_sys::GUID {
    Data1: 0x75B54759,
    Data2: 0xB6F1,
    Data3: 0x49C2,
    Data4: [0xA2, 0x09, 0x68, 0x54, 0x96, 0xBD, 0x70, 0xC0],
};

#[allow(dead_code)]
const PAYLOAD_NVAPI_GUID: detours_sys::GUID = detours_sys::GUID {
    Data1: 0xc01fff7f,
    Data2: 0x310d,
    Data3: 0x4ac4,
    Data4: [0xAF, 0x47, 0x11, 0x4d, 0x9a, 0xef, 0x7a, 47],
};

#[allow(dead_code)]
const PAYLOAD_NVOPTIX_GUID: detours_sys::GUID = detours_sys::GUID {
    Data1: 0x629e45e5,
    Data2: 0xf4d1,
    Data3: 0x4649,
    Data4: [0x9d, 0xb4, 0x7e, 0x17, 0xf6, 0x74, 0xc2, 0xad],
};
```
Are they set (fixed) values or random/generated values?
If they are fixed values, where do they come from?
I searched out, but I couldn't find anything helpful.

**Comment by vosen** [comment, marked as answer]
Those GUIDs are keys used by Microsoft Detours library. When ZLUDA launcher (zluda.exe) launches a process it injects the zluda redirector dll and paths to CUDA dlls. The GUIDs are keys used to write and read path information.
For example when you execute
```
c:\foo\zluda.exe -- c:\bar\app.exe
```
then injector inside `app.exe` knows that using key `PAYLOAD_NVCUDA_GUID` it can get path to `nvcuda.dll` that is `c:\foo\nvcuda.dll`.
See Detours wiki here: https://github.com/microsoft/Detours/wiki/DetourBinarySetPayload and here: https://github.com/microsoft/Detours/wiki/DetourFindPayload
